neothesia/src/scene/playing_scene/midi_player.rs

```diff
@@ -198,69 +252,201 @@ pub enum MidiEventSource {
+                if let Some(item) = self
+                    .user_pressed_recently
+                    .iter_mut()
+                    .find(|item| item.note_id == note_id)
+                {
+                    // already exists, update timestamp
+                    item.timestamp = timestamp;
+                    self.user_stats.wrong_notes += 1;
+                } else {
+                    // Not found, push a new UserPress
+                    self.user_pressed_recently.push_back(UserPress {
+                        timestamp,
+                        note_id,
+                        occurrence: *occurrence,
+                        time_key_up: timestamp,
+                    });
+                }
+            }
+        } else {
+            // Update user_notes log time_key_up
+            if let Some(item) = self
+                .user_notes
+                .iter_mut()
+                .rev()
+                .find(|item| item.note_id == note_id && item.occurrence == *occurrence)
+            {
+                item.time_key_up = Instant::now();
+            }
         }
     }
 
     fn file_press_key(&mut self, note_id: u8, active: bool) {
+        let occurrence = self.occurrence.entry(note_id).or_insert(0);
+        let timestamp = Instant::now();
         if active {
-            if let Some((id, _)) = self
+            *occurrence += 1;
+
+            // Check if note got pressed earlier 500ms (user_pressed_recently)
+            if let Some(item) = self
                 .user_pressed_recently
                 .iter()
-                .enumerate()
-                .find(|(_, item)| item.note_id == note_id)
+                .find(|item| item.note_id == note_id)
             {
-                self.user_pressed_recently.remove(id);
+                // Note was pressed earlier, remove it from user_pressed_recently
+                self.user_stats.notes_hit += 1;
+
+                // log user_note by user_pressed_recently.timestamp as keydown value, update occurence
+                self.user_notes.push_back(UserPress {
+                    timestamp: item.timestamp,
+                    note_id,
+                    occurrence: *occurrence,
+                    time_key_up: item.time_key_up,
+                });
+                self.user_pressed_recently
+                    .retain(|item| item.note_id != note_id);
             } else {
-                self.required_notes.insert(note_id);
+                // Player never pressed that note, let it reach required_notes, check if note_id already exists in required_notes,  update timestamp else push.
+                // Catch possible clone-note velocity overlay, update the new occurence and exit the function
+                if let Some((_id, item)) =
+                    self.file_notes.iter_mut().enumerate().find(|(_, item)| {
+                        item.note_id == note_id && item.time_key_up == item.timestamp
```

Comment:
Is this a check for notes that have not yet ended?
Sounds quite weird to loop over every note in the song just to find the inproggres notes.